### ISSUE TITLE
Support HTTP2 probes over cleartext (h2c)

### ISSUE THREAD

**Issue by grepwood** [kubernetes/kubernetes#125599]
I'd like to request that this use case becomes supported. At work, we are running a cluster behind a massive HTTP load balancer, so we have no real reason to support TLS in our microservices and ingresses. The load balancer takes care of that.

I can't really say that I like what happened in the original issue https://github.com/kubernetes/kubernetes/issues/97330...

@lol768 is another user that has this specific use case in mind.

Now, because this use case was not accommodated in the past, I now have to introduce a really frankensteinian network setup into my namespace.

----- BEGIN EDIT 1 -----
The reason why I've even introduced HTTP/2 support into my microservices is because there are clear advantages of HTTP/2 over HTTP/1.1 in the use case of the project I work for, that have to do with efficiency and performance.

----- END EDIT 1 -----

I now have to make sure that every web service has a second listening port where it only emits HTTP/1.1 so that Kubernetes will be able to issue probes into it. And to make this port a probe-only port that doesn't get exposed outside, I need to make sure that my ingresses will never expose it, or anyone from the team never dares to expose it. And to make sure that the probe-only port doesn't leak any sensitive data, I have to make sure that it returns only some kind of generic HTTP response to the probes.

At this point you're probably thinking why won't I just get a TCP probe to knock on the port where we serve HTTP/2. This will work of course, but it's naive. It misses the point. If this was a TCP service that doesn't serve HTTP, it would make perfect sense, but it isn't. HTTP is not just TCP 80 or 8080 or what have you, but it's a probe with extra context - the return is expected to be a HTTP response, so that we know the microservice isn't just merely accepting connections, it's also sending coherent responses.

I hope this is explanation enough and that its merits are reasonable.

**Comment by mauri870** [comment]
From the information of both issues, I see this as a totally valid use case. 
It is common to have HTTP servers that run in cleartext mode and do SSL termination on some upper load balancer layer.

What changes are you actually proposing @grepwood?

Should we add counterparts to the `httpGet` and `grpc` type probes for http2, perhaps even separated in `http2Get` and `h2cGet`?

**Comment by grepwood** [comment]
@mauri870 yes please. All the solutions you've listed are good from my point of view, so I think it's up to developers to decide which one would give them least amount of issues now and later.

**Comment by mauri870** [comment]
I was looking a bit into the current implementation, looks like setting httpGet with scheme set to HTTPS already works with HTTP2 over tls.

Playing devil's advocate, I wonder if it is worth the effort to implement an `h2cGet` considering that one can just register a HTTP1.1 route `/healthz` just for probing. Saying this because we are looking at supporting all options that httpGet supports today (minus scheme) and also a whole new implementation for h2c.

**Comment by mauri870** [comment]
Some more details I found. To make a http client an h2c client we just have to change the transport:  

https://github.com/thrawn01/h2c-golang-example/blob/931670014b5a6d7b7fa5fa8482dc1d98ab71eadc/cmd/client/main.go#L51-L59

That seems to plug nicely with our http prober:

https://github.com/kubernetes/kubernetes/blob/e6ea4f0538285f8bdf3a493088be02610afa5c34/pkg/probe/http/http.go#L46-L55

Now it does not seem so complex to support `h2c`, but I might have overlooked something.

**Comment by mauri870** [comment]
/retitle Support HTTP2 probes over cleartext (h2c)

**Comment by lol768** [comment]
Thank you for reviving this one @grepwood 

You've set out all of the motivation here quite eloquently.

As @mauri870 has observed, it shouldn't be very complex to support this at all.

> Playing devil's advocate, I wonder if it is worth the effort to implement an h2cGet considering that one can just register a HTTP1.1 route /healthz just for probing

@mauri870 with the framework I use (Kestrel/ASP.NET) you _cannot_ just do HTTP/1.1 and unencrypted HTTP/2 at the same time on the same port because it _needs_ ALPN to be either to do protocol negotiation, which relies on TLS.

I have to do something like this today:

```
builder.WebHost.ConfigureKestrel(options =>
{
    // port 8282 for HTTP/2 (unencrypted)
    options.Listen(IPAddress.Any, 8282, listenOptions =>
    {
        listenOptions.Protocols = HttpProtocols.Http2;
    });
    // port 80 for k8s probe which is too stupid to understand clear HTTP/2 - #97330
    options.Listen(IPAddress.Any, 80, listenOptions =>
    {
        listenOptions.Protocols = HttpProtocols.Http1;
    });
});
```

See also https://github.com/grpc/grpc-dotnet/issues/979#issuecomment-656966382

**Comment by grepwood** [comment]
> Playing devil's advocate, I wonder if it is worth the effort to implement an `h2cGet` considering that one can just register a HTTP1.1 route `/healthz` just for probing. Saying this because we are looking at supporting all options that httpGet supports today (minus scheme) and also a whole new implementation for h2c.

That's exactly what I'm going to do and I don't think this is a good test. Such a probe will only test if we can emit HTTP/1.1 responses specifically designed to return healthchecks. It's not gonna test a generic endpoint so that we know the microservice returns HTTP responses relevant to its job.

**Comment by grepwood** [comment]
> That's exactly what I'm going to do and I don't think this is a good test. Such a probe will only test if we can emit HTTP/1.1 responses specifically designed to return healthchecks. It's not gonna test a generic endpoint so that we know the microservice returns HTTP responses relevant to its job.

I did exactly this, and it works. It's slightly more messy than if we supported HTTP/1.1 and h2c on the same socket, but it works.

**Comment by mauri870** [comment]
Apologies for letting this go stale.

I'm not sure if an h2c-type probe would be better suited as an option within httpGet or if it should be implemented as a separate probe altogether.

**Comment by 75lb** [comment]
My application is HTTP2-only (since it requires persistent connections, push streams etc) and doesn't require TLS as all comms are end-to-end encrypted (for privacy reasons - with TLS, it's still possible for the server to read all traffic).

So unfortunately for me, all k8s HTTP probes fail until this issue is fixed or I develop a workaround. I chose to abandon k8s entirely and just manage the VM myself.

**Comment by grepwood** [comment]
I had to really stretch what is possible in my workplace and managed to get nginx 1.26.0, where we can listen for HTTP and HTTP/2 traffic on the same socket. Still we would prefer to take advantage of the newer protocol.

**Comment by greguz** [comment]
Up this thing 🥸

**Comment by grepwood** [comment]
Yep, this is much desired :)

**Comment by lol768** [comment]
/remove-lifecycle stale

Stale bots genuinely show nothing but contempt for the people who invest time in writing up feature requests and bug reports.

**Comment by C0DK** [comment]
https://github.com/kubernetes/kubernetes/issues/125599#issuecomment-2608975840

> After 90d of inactivity, lifecycle/stale is applied

Just to make sure this doesn't go stale, I'd like to re-emphasis interest in this, as the current fix is hacking the application similar to what @lol768

**Comment by greguz** [comment]
> Apologies for letting this go stale.
> 
> I'm not sure if an h2c-type probe would be better suited as an option within httpGet or if it should be implemented as a separate probe altogether.

I propose to add a different `scheme` for H2C probes, like `scheme: HTTP2_CLEARTEXT` or something similar.

That's because HTTPS has ALPN, but HTTP/1.1 and HTTP/2 are just different protocols, so (for me) makes sense to have different schemes.

**Comment by lol768** [comment]
Pointless bump to prevent addition of `stale` label in a few days.

**Comment by grepwood** [comment]
At this point I'm considering actually doing this. Kubernetes is written in Go. I never wrote a single line of Go, but it's never too late to do so.

**Comment by lol768** [comment]
Hopefully they're less bad at triaging PRs?

**Comment by aojea** [comment]
> Hopefully they're less bad at triaging PRs?

@lol768 this is an OSS project and everyone can work on things, but there is no KEP or proposal filed AFAIK https://github.com/kubernetes/enhancements/tree/master so I do not understand the sarcasm

**Comment by lol768** [comment]
I think it's quite a reasonable concern to raise - will a PR get reviewed or simply left and auto-closed as stale?

More widely I think the advent of these bots that auto-close well researched / written issues to be in really poor taste. Either it's a valid issue and can stay open until somebody has resource to look into addressing it, or it's something the project doesn't want and should be closed as "Not planned". It's disrespectful to everyone who's invested time looking into this to say "actually, fuck you - we're going to automatically close it with a bot".

It is very tempting to write an anti-anti-stale bot to automatically keep issues alive!

**Comment by aojea** [comment]
> More widely I think the advent of these bots that auto-close well researched / written issues to be in really poor taste. Either it's a valid issue and can stay open until somebody has resource to look into addressing it, or it's something the project doesn't want and should be closed as "Not planned". It's disrespectful to everyone who's invested time looking into this to say "actually, fuck you - we're going to automatically close it with a bot".

@lol768 I do not think nobody disagree with the problem statement, but unfortunately there is no perfect solution that keeps everyone happy, see previous discussions https://github.com/kubernetes/kubernetes/issues/103151

**Comment by grepwood** [comment]
>I do not understand the sarcasm

See [innoextract](https://github.com/dscharrer/innoextract) and [piskel](https://github.com/piskelapp/piskel). These are very hot OSS projects with many old issues and PRs - [some old enough that if they were children, they would be looking at going to secondary school in a year or two](https://github.com/dscharrer/innoextract/pull/53). [Some are old enough that if they were human, they would be going through puberty](https://github.com/dscharrer/innoextract/issues/9). Unlike Kubernetes, these projects own their state of disarray rather than pretend everything is hunky dory while a bot shoves all the skeletons back in the closet.

Edit: also, this issue is a followup to another issue that was swept under the rug. The original merit of https://github.com/kubernetes/kubernetes/issues/97330 continued by this issue, is old enough to be graduating from kindergarten in a year or two.

**Comment by lol768** [comment]
> If you want, I can draft an even snappier, casual version that’s just a line or two

_Really?_ 🙄

**Comment by mauri870** [comment]
Removing me from the assignees since I clearly don't have capacity to work on this right now. Thanks for offering to work on this issue!

/unassign @mauri870
/assign @amritansh1502

**Comment by SergeyKanzhelev** [comment]
This is being tracked as a KEP now: https://github.com/kubernetes/enhancements/issues/5999

@amritansh1502 is there any reason to keep this bug open?

**Comment by grepwood** [comment]
@SergeyKanzhelev yes, to keep the stakeholders informed such as myself and @lol768. Shouldn't be closed without fixing.

**Comment by lol768** [comment]
Can we at least stop it being automatically closed now that things are progressing?

**Comment by SergeyKanzhelev** [comment]
/kind feature
/priority backlog
/triage accepted

I do not mind having it open. It is just easier to track it as a KEP now. 

As stakeholders, drop comments to PR: https://github.com/kubernetes/kubernetes/pull/139429 if you see anything wrong and give feedback while it is in alpha (all goes well - it will be in alpha in 1.37)
